src/Microsoft.Graph.Core/Tasks/LargeFileUploadTask.cs

```diff
@@ -0,0 +1,258 @@
+                        {
+                            // Succeeded previously, but nothing to return right now
+                            return new UploadResult<T>();
+                        }
+                        else
+                        {
+                            throw;
+                        }
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Get the series of requests needed to complete the upload session. Call <see cref="UpdateSessionStatusAsync"/>
+        /// first to update the internal session information.
+        /// </summary>
+        /// <param name="options">Options to be applied to each request.</param>
+        /// <returns>All requests currently needed to complete the upload session.</returns>
+        public virtual IEnumerable<UploadSliceRequest<T>> GetUploadSliceRequests(IEnumerable<Option> options = null)
+        {
+            foreach (var (item1, item2) in this._rangesRemaining)
+            {
+                var currentRangeBegins = item1;
+
+                while (currentRangeBegins <= item2)
+                {
+                    var nextSliceSize = NextSliceSize(currentRangeBegins, item2);
+                    var uploadRequest = new UploadSliceRequest<T>(
+                        this.Session.UploadUrl,
+                        this._client,
+                        options,
+                        currentRangeBegins,
+                        currentRangeBegins + nextSliceSize - 1,
+                        this.TotalUploadLength);
+
+                    yield return uploadRequest;
+
+                    currentRangeBegins += nextSliceSize;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Upload the whole session.
+        /// </summary>
+        /// <param name="maxTries">Number of times to retry entire session before giving up.</param>
+        /// <param name="options">Query and header option name value pairs for the request.</param>
+        /// <param name="progress">IProgress object to monitor the progress of the upload.</param>
+        /// <returns>Item information returned by server.</returns>
+        public async Task<UploadResult<T>> UploadAsync(IProgress<long> progress = null, int maxTries = 3, IEnumerable<Option> options = null)
+        {
+            var uploadTries = 0;
+            var trackedExceptions = new List<Exception>();
+
+            while (uploadTries < maxTries)
+            {
+                var sliceRequests = this.GetUploadSliceRequests(options);
+
+                foreach (var request in sliceRequests)
+                {
+                    var uploadResult = await this.UploadSliceAsync(request, trackedExceptions).ConfigureAwait(false);
+                    
+                    progress?.Report(request.RangeBegin);//report the progress of upload
+
+                    if (uploadResult.UploadSucceeded)
+                    {
+                        return uploadResult;
```

Comment:
Should we delete the upload session for the user if the upload succeeded?

Reply:
This is not necessary. On successful upload, the session will be deleted automatically from the server end.